autofit/graphical/mean_field.py

```diff
@@ -1,291 +1,650 @@
+
+        if axis is not None:
+            raise NotImplementedError(
+                "FactorApproximation.func_jacobian has not implemeted "
+                f"axis={axis}, try axis=None")
+
+        if variables is None:
+            fixed_variables = set(
+                v for v, m in self.model_dist.items() 
+                if isinstance(m, FixedMessage))
+            variables = self.factor.variables - fixed_variables
+
+        fval, fjac = self.factor.func_jacobian(
+            variable_dict, variables, axis=axis, 
+            _calc_deterministic=_calc_deterministic)
+
+        values = {**variable_dict, **fval.deterministic_values}
+        var_sizes = {v: np.size(x) for v, x in values.items()}
+        var_shapes = {v: np.shape(x) for v, x in values.items()}
+        log_cavity, grad_cavity = self.cavity_dist.logpdf_gradient(
+            values, axis=axis)
+
+        logl = fval + log_cavity
+
+        for v in fjac:
+            fjac[v] += grad_cavity[v]
+
+        # Update gradients using jacobians of deterministic variables
+        # TODO: Should add logic to account for pullbacks for 
+        #       AD frameworks e.g. Zygote.jl
+        for var, grad in fjac.items():
+            for det, jac in grad.deterministic_values.items():
+                det_grad = grad_cavity[det].ravel()
+                g = jac.reshape(var_sizes[det], var_sizes[var])
+                fjac[var] += det_grad.dot(g).reshape(var_shapes[var])
+        
+        return logl, fjac
+
+    def project_mean_field(
+            self, 
+            model_dist: MeanField, 
+            delta: float = 1.,
+            status: Optional[Status] = None,
+    ) -> "FactorApprox":
+        success, messages = Status() if status is None else status
+
+        factor_dist = (model_dist / self.cavity_dist)
+        if delta < 1:
+            log_norm = factor_dist.log_norm
+            factor_dist = (
+                factor_dist**delta * self.factor_dist**(1-delta))
+            factor_dist.log_norm = (
+                delta * log_norm + (1 - delta) *  self.factor_dist.log_norm)
+
+        if not factor_dist.is_valid:
+            success = False
+            messages += (
+                f"model projection for {self} is invalid",)
+            factor_dist = factor_dist.update_invalid(self.factor_dist)
+        
+        # variable_norm = self.cavity_dist.log_normalisation(factor_dist)
+        # factor_dist.log_norm = np.sum(factor_dist.log_norm) - variable_norm
+
+        new_approx = FactorApproximation(
+            self.factor,
+            self.cavity_dist, 
+            factor_dist=factor_dist,
+            model_dist=model_dist,
+        )
+        return new_approx, Status(success, messages)
+
+    project = project_mean_field
+
+    def __repr__(self):
+        # TODO make this nicer
+        return f"{type(self).__name__}({self.factor}, ...)"
+
+    
+class MeanFieldApproximation:
+    '''
+    TODO: delete this class
```

Comment:
Delete this class?

Reply:
Yeah I need to update the whole test suite to do that...